### ISSUE TITLE
at_activate in C

### ISSUE THREAD

**Issue by xavierchanth** [atsign-foundation/noports#1325]
Dependent on atsign-foundation/at_c#176 
```[tasklist]
### Tasks
- [ ] https://github.com/atsign-foundation/at_c/issues/176
- [ ] atsign-foundation/at_c#74 
```

**Comment by JeremyTubongbanua** [comment]
Where would this project live @XavierChanth ?

Would it be in `examples/` ?

**Comment by cpswan** [comment]
I'm guessing here, as @XavierChanth hasn't explained *why*...

at_activate is a crucial component of the customer journey, and if their device needs C sshnpd then it likely also needs a smaller and more portable at_activate.

So not an example really, but also not part of the SDK. Maybe we need a utils directory, or maybe we need to put it in the noports repo alongside the daemon.

**Comment by JeremyTubongbanua** [comment]
One thing we could also consider is making an entirely new at_tools repo made in C, like `at_tools_c` that would house generic Atsign tools like `repl`, `at_talk`, `atcat` & `attee` (https://github.com/atsign-foundation/at_demos/issues/309), and also our `at_activate`. 

`repl` and `at_talk` currently live in our `examples/` directory.

**Comment by xavierchanth** [comment]
The primary object for creating at_activate is to make it easier to do APKAM / atSign onboarding with the C daemon. I am hesitant to put it in examples, but maybe a `bin/` directory at the root of this repo. repl & at_talk can be moved into this new directory.

**Comment by cconstab** [comment]
This is now something that needs to float to the top of the ToDo's.. It will enable a number of customers to automate sshnpd and its keys on IoT devices without logging into the hardware. 

**MVP**
at_activate with APKAM creation of atKeys

**Comment by srieteja** [comment]
I will take this up this sprint @cconstab

**Comment by cconstab** [comment]
APKAM functionality is the target.. TY

**Comment by JeremyTubongbanua** [comment]
We would like to move this ticket to [NoPorts](https://github.com/atsign-foundation/noports). We would like to have at_activate be within the NoPorts repo to make things simple for our CI when we package NoPorts zips and binaries

Would it be possible to change this ticket's repo? I don't know how to do that

**Comment by srieteja** [comment]
Carrying forward to PR97 to implement key/keypair generation and exchange these keys with the at_server.

**Comment by srieteja** [comment]
@JeremyTubongbanua, @XavierChanth and I have completed the implementation of dependencies required to implement at_activate. But there is one bug which was interrupting my progress (MbedTLS read returns a connection timed out err), @JeremyTubongbanua is going to take a look at this bug today as I'm not very familiar with MbedTLS and it's eating up a lot of my time.

In the meantime, I have started working on at_activate itself. Current ETA to complete this would be Tuesday (at the latest).

**Comment by srieteja** [comment]
Blocked by https://github.com/atsign-foundation/at_c/issues/426

**Comment by srieteja** [comment]
https://github.com/atsign-foundation/at_c/issues/426 has been resolved. See ticket for explanation.

**Comment by cconstab** [comment]
Hows this ticket progressing.. looks like it is close but hard to tell .. Itching to test :-)

**Comment by srieteja** [comment]
All the implementation is complete @cconstab. Trying to resolve all the obvious bugs. You should have a version to test in a day or two.

**Comment by srieteja** [comment]
@purnimavenkatasubbu found a bug in an atserver that was activated with C atactivate. The server throws a publickey not found error when updating data. Looking into this. 

Pending items:
1) above mentioned bug
2) unit tests (if there are any unit testable elements)

**Comment by srieteja** [comment]
The at_activate PR has been merged. The implementation to create enrollments with OTP is being tracked through https://github.com/atsign-foundation/at_c/issues/451.

**Comment by srieteja** [comment]
Closing this ticket as the implementation to create new enrollments using OTPs has been merged to trunk through: https://github.com/atsign-foundation/at_c/issues/451
